src/pulse3D/nb_peak_detection.py

```diff
@@ -0,0 +1,209 @@
+        max_frequency: The maximum frequency (Hz) at which a peak can occur. Specifically, this is used to calculate the minimum required distance between adjacent peaks.
+            For example, if the value given is 1, then at most peaks will occur at 1Hz. In other words, the peaks will be no closer than 1 second.
+            If not specified, the sampling frequency is used instead, which means that every point can be considered a peak.
+            If a value is given that exceeds the sampling frequency, the sampling frequency is used instead.
+        valley_search_duration: The duration of time in seconds prior to a peak in which to search for a valley.
+            If this window includes a previous peak then for that peak the window will automatically be shortened
+        upslope_duration: The min duration of time in seconds through which the waveform values must continuously rise in order to be considered an upslope.
+        upslope_noise_allowance_duration: The max duration of time in seconds in the upslope in which there is an amplitude decrease which can be tolerated within a single upslope.
+
+    Returns:
+        A tuple containing an of the indices of the peaks and an array of the indices of valleys
+    """
+    time_axis, waveform = tissue_data
+
+    # TODO split into subfunctions, one for finding peaks and the other for finding valleys
+
+    # extract sample frequency from time_axis (assumes sampling freq is constant)
+    sample_freq = 1 / (time_axis[1] - time_axis[0])
+
+    if max_frequency:
+        # if max freq is greater than the sampling freq, use sampling freq instead
+        max_frequency = min(max_frequency, sample_freq)
+    else:
+        # if no max freq given, use sampling freq
+        max_frequency = sample_freq
+
+    # set estimate of peak to peak noise amplitude is 10uN for average recording
+    default_noise = 10
+    default_prom = 5
+
+    # find peaks with this estimated amplitude
+    peaks, _ = signal.find_peaks(waveform, prominence=default_prom * default_noise)
+
+    # if first attempt finds no peaks as they are too small, retry with smaller prominence
+    # this approach should return a list of peak indices even if no true peaks exist as it will terminate at a prominence of 1.
+    correction_factor = 1
+    while len(peaks) == 0 and correction_factor <= default_prom:
+        peaks, _ = signal.find_peaks(waveform, prominence=(default_prom - correction_factor) * default_noise)
+
+        correction_factor += 1
+
+    if (num_peaks := len(peaks)) < MIN_NUMBER_PEAKS:
+        raise TooFewPeaksDetectedError(
+            f"A minimum of {MIN_NUMBER_PEAKS} peaks is required to extract twitch metrics, however only {num_peaks} peak(s) were detected."
+        )
+
+    # use peaks to extract waveform segments from which noise data can be extracted - control over this could be given to the user if required
+    segment_size = 10
```

Comment:
if this is a const value I'd make it all caps

Reply:
nvm mind, I see it's updated below

Reply:
although it might be nice to make `segment_size` a param that has a default value of `10` or make a const `DEFAULT_SEGMENT_SIZE = 10` that is assigned to this variable